src/rhsm/connection.py

```diff
@@ -343,54 +340,95 @@ def _request(self, request_type, handler, body=None, headers=None):
-            "headers": dict(response.getheaders())}
+    def _get_ent_cert_key_list(self):
+        """
+        Create list of cert-key pairs that can be used for connection with CDN
+        :return: List of tuples containing certificate and private key
+        """
 
-        return result
+        ent_cert_key_pairs = []
 
-    def _load_ca_certificates(self, context):
-        cert_path = ''
+        for cert_file in os.listdir(self.ent_dir):
+            if cert_file.endswith(".pem") and not cert_file.endswith("-key.pem"):
+                cert_path = os.path.join(self.ent_dir, cert_file)
+                key_path = os.path.join(self.ent_dir, "%s-key.pem" % cert_file.split('.', 1)[0])
+                ent_cert_key_pairs.append((cert_path, key_path))
+
+        return ent_cert_key_pairs
+
+    def _load_ca_certificate(self, context, cert_path, key_path):
+        """
+        Try to load certificate
+        :param context:
+        :param cert_path:
+        :param key_path:
+        :return:
+        """
+        log.debug("Loading CA certificate: '%s'" % cert_path)
         try:
-            for cert_file in os.listdir(self.ent_dir):
-                if cert_file.endswith(".pem") and not cert_file.endswith("-key.pem"):
-                    cert_path = os.path.join(self.ent_dir, cert_file)
-                    key_path = os.path.join(self.ent_dir, "%s-key.pem" % cert_file.split('.', 1)[0])
-                    log.debug("Loading CA certificate: '%s'" % cert_path)
-
-                    # FIXME: reenable res =
-                    context.load_verify_locations(cert_path)
-                    context.load_cert_chain(cert_path, key_path)
-                    # if res == 0:
-                    #     raise BadCertificateException(cert_path)
+            context.load_verify_locations(cert_path)
+            context.load_cert_chain(cert_path, key_path)
         except ssl.SSLError as e:
             raise BadCertificateException(cert_path)
         except OSError as e:
             raise ConnectionSetupException(e.strerror)
 
+    def _request(self, request_type, handler, body="", headers=None, ent_cert_key_pairs=None):
+        # See note in Restlib._request
+        context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+
+        # Disable SSLv2 and SSLv3 support to avoid poodles.
+        context.options = ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3
+
+        final_headers = {"Host": "%s:%s" % (normalized_host(self.host), self.ssl_port),
+                         "Content-Length": "0",
+                         "User-Agent": self.user_agent}
+        if headers:
+            final_headers.update(headers)
+
+        # When no list of entitlement certificates is provided, then try to use all
+        # entitlement certificates installed on the system
+        if ent_cert_key_pairs is None or len(ent_cert_key_pairs) == 0:
+            ent_cert_key_pairs = self._get_ent_cert_key_list()
+
+        # Just anything else than 200
+        result = {"status": 123456}
```

Comment:
I'd rather we do something like

```
for cert_path, key_path...
    ...
    if response.status == 200:
        return result
raise SomeAppropriateException
```

Because not having a cert_path or key_path and thus not being able to initiate a connection at all is an exceptional case.  I worry that returning a dummy value creates the potential for issues where we assume in higher layers of code that `result` is something valid when it is not.

Reply:
Sounds good.